### PR TITLE
NeMo 2.0 vLLM export

### PR DESCRIPTION
**Pull request by Laplasjan107**
# What does this PR do ?

vLLM export for NeMo 2.0 models

**Collection**: llm

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
The vLLM export needs to be run in the virtual environment, installed inside the NeMo container:
``` bash
source /opt/venv/bin/activate
```

Sample vLLM script for export:
```python
from nemo.export.vllm_exporter import vLLMExporter
import os, os.path
import argparse

def get_args():
    parser = argparse.ArgumentParser(
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
        description="vLLM export argument parser",
    )
    parser.add_argument("--nc", type=str)
    parser.add_argument("--tps", type=int)
    parser.add_argument("--mt", type=str)

    args = parser.parse_args()
    args.export_path = f"/tmp/vllm_{args.mt}_tps{args.tps}"
    return args


args = get_args()
if not os.path.exists(args.export_path):
   os.mkdir(args.export_path)

exporter = vLLMExporter()
exporter.export(
   nemo_checkpoint=args.nc,
   model_dir=args.export_path,
   model_type=args.mt,
   tensor_parallel_size=args.tps,
)

output = exporter.forward(["What is the best city in the world?"], max_output_len=50, top_k=1, top_p=0.0, temperature=1.0)
print("output: ", output)
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [X] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/export/sentencepiece_tokenizer.py [hunk 1]
```diff
@@ -29,11 +29,28 @@ class SentencePieceTokenizer:
...
     def __init__(
-        self, model_path: str, special_tokens: Optional[Union[Dict[str, str], List[str]]] = None, legacy: bool = False
+        self,
+        model_path: Optional[str] = None,
+        special_tokens: Optional[Union[Dict[str, str], List[str]]] = None,
+        legacy: bool = False,
+        tokenizer: Optional[sentencepiece.SentencePieceProcessor] = None,
     ):
+        if tokenizer is not None:
+            self.tokenizer = tokenizer
```
Comment (janekl) [inline nemo/export/sentencepiece_tokenizer.py:43]: could this go into some auxiliary method as it's also kind of repeated in L59-70?

File: nemo/export/vllm/model_config.py [hunk 2]
```diff
@@ -101,26 +104,55 @@ def __init__(
...
-
+        if is_nemo2_checkpoint(nemo_checkpoint):
+            from nemo.lightning.io import load_context
+
+            nemo_checkpoint: Path = Path(nemo_checkpoint)
+            self.nemo_model_config: dict = yaml.load(
+                (nemo_checkpoint / "context/model.yaml").open('r'), Loader=yaml.SafeLoader
+            )
+            hf_args = {}
+            for hf_arg, nemo_arg in hf_to_nemo_dict.items():
```
Comment (janekl) [inline nemo/export/vllm/model_config.py:118]: repeated in L144-153: maybe helper?

File: nemo/export/vllm/engine.py [hunk 3]
```diff
@@ -25,15 +27,46 @@
 LOGGER = logging.getLogger("NeMo")
 
 
+class VLLMTokenizerGroup(TokenizerGroup):
```
Comment (oyilmaz-nvidia) [inline nemo/export/vllm/engine.py:30]: A minor thing but can we rename this to `vLLMTokenizerGroup`

**Review by oyilmaz-nvidia — CHANGES_REQUESTED** [review]
Were you able to test this with a nemo 2 ckpt?

**Comment by janekl** [comment]
> Were you able to test this with a nemo 2 ckpt?

Yep, that's the very point here. Model types tested are Llama, Mixtral, Gemma, StarCoder2 to keep vLLM supported models parity with 24.07 release https://docs.nvidia.com/nemo-framework/user-guide/24.07/deployment/llm/optimized/index.html

File: nemo/export/vllm/model_loader.py [hunk 4]
```diff
@@ -15,23 +15,58 @@
...
+from torch.distributed.checkpoint.metadata import TensorStorageMetadata
+from torch.distributed.checkpoint.state_dict_loader import load_state_dict
 from vllm.config import CacheConfig, DeviceConfig, LoRAConfig, ModelConfig, ParallelConfig, SchedulerConfig
 from vllm.model_executor.model_loader.loader import BaseModelLoader, _initialize_model
 from vllm.model_executor.model_loader.utils import set_default_torch_dtype
 
 from nemo.export.tarutils import TarPath, ZarrPathStore
 from nemo.export.vllm.model_config import NemoModelConfig
 
+from .utils import is_nemo2_checkpoint
```
Comment (janekl) [inline nemo/export/vllm/model_loader.py:36]: Let's please keep one convention for imports so I think full-path like `from nemo.export.vllm.utils import ...`

**Review by janekl — APPROVED** [review]

**Review by ko3n1g — APPROVED** [review]

**Review by oyilmaz-nvidia — APPROVED** [review]
LGTM, thanks!
